### PR TITLE
Added some simple AppBar examples to the sample catalog

### PR DESCRIPTION
**Pull request by HansMuller**
Simple AppBar samples. The idea here is to establish a pattern for sample catalog apps. 

The comment near the top of each file is meta-information  that I hope we'll parse and use to create links in  the widget catalog and the API doc.

I think we'll also want to automatically generate screenshots of the sample apps in action.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @sethladd @InMatrix 
I would love to get Seth and Tao's feedback on this, since it's mainly about helping people learn about the features. What works, what doesn't?

**Comment by InMatrix** [comment]
I'd love to start with some rough sketches of what the user will see and what she will be able to do instead of code. Can we chat after I get back in office on Friday?

**Comment by HansMuller** [comment]
Here's a rough sketch of how I think a sample catalog might integrate with the existing widget catalog and API docs.

The widget catalog and the API doc would just provide a boilerplate "samples" link that pointed to a catalog entry for the corresponding widget.

Something roughly like this:

catalog

- We add flutter.io/samples. It's the "sample catalog" and it would be mostly machine generated.
- The catalog entries would be generated based the "Sample Catalog" comment in the source code for each sample app.
- The screenshots would be generated by running the apps.
- Ideally each screenshot would effectively be a (short, like < 5 seconds) video that played when you rolled over it.
- Whatever "beyond ideally" is. The realm of fantasy. Rolling over widget names in the descriptions below each entry would highlight the widget's boundary in the screenshot.

**Comment by InMatrix** [comment]
Thanks for the visual examples. I think building the infrastructure to generate easy-to-run samples and screenshots from annotated code is great. The ability to identify widgets boundary in the screenshots and perhaps even annotate them with widgets' names is a really nice fantasy to have. Maybe this won't be hard once we have a visual inspector.

I'm all for experimenting with the sample generation infrastructure now, but we need to do more planning on content, i.e., what samples we actually need to provide. One question is the granularity of the samples. I'd recommend us take a look at the [Atomic Design Methodology](http://atomicdesign.bradfrost.com/chapter-2/), where the author provides a framework to group UI elements. Mapping that metaphor to Flutter, the widget catalog is an "atomic" catalog, since it introduces individual widgets. I think the sample catalog needs to distinguish itself by hosting samples at the levels of molecules and organisms. For example, an infinite scrolling UI pattern is a good candidate for the sample catalog, while a ListView belongs to the widget catalog. For the atomics (aka, widgets) in Flutter, we should provide sample code and screenshots in their API doc since this is where developers are accustomed to.

**Comment by sethladd** [comment]
As you're thinking about scope, please include how we'll cross-link these samples into our Widget Catalog. It should be useful to bounce between Sample Catalog and Widget Catalog.

I'd also love to see if/how we can display the source code for the sample, in the browser. Maybe it's just linking over to GitHub? If there's concern about bouncing the user back and forth, we might want to find a way to inline the source on the flutter.io/samples/____ pages.

Thanks for working on this!

File: examples/catalog/lib/app_bar_bottom.dart [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+import 'package:flutter/material.dart';
+
+/// Sample Catalog: AppBar with a custom bottom widget.
+///
+/// The bottom widget's TabPageSelector displays the relative position of the
+/// selected choice. The arrow buttons in the toolbar part of the app bar
+/// select the previous or the next choice.
+///
+/// Sample classes: [AppBar], [TabController], [TabPageSelector], [Scaffold], [TabBarView].
```
Comment (sethladd) [inline examples/catalog/lib/app_bar_bottom.dart]: Love this. Thanks!

File: examples/catalog/lib/app_bar_bottom.dart [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+///
+/// See also:
+///  * The "Components-Tabs" section of the material design specification:
+///    <https://material.io/guidelines/components/tabs.html>
+
+class Choice {
```
Comment (sethladd) [inline examples/catalog/lib/app_bar_bottom.dart]: Did we think about how to more clearly mark the code in the sample that is specific to the sample (like, for instance, this Choice class), and the code that is generic and meant to be copy-pasted by the developer? I found myself scanning through this code, looking for the UI-specific part (the part about the AppBar, TabController, etc). I'm immediately presented with this "app state" code, so I'm left wondering "do I need this?"

Not sure how to do it, but I'd love to clearly mark/separate/draw-the-eye to the generic parts of this sample. The part I'm going to copy and paste.

File: examples/catalog/lib/app_bar_bottom.dart [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    _tabController = new TabController(vsync: this, length: choices.length);
+  }
+
+  @override
+  void dispose() {
+    _tabController.dispose();
+    super.dispose();
+  }
+
+  void _nextPage(int delta) {
```
Comment (sethladd) [inline examples/catalog/lib/app_bar_bottom.dart]: here's a good example: is this required? Do I need to copy and paste this? I encounter this function, before it's used, so I don't have a lot of context.

**Comment by HansMuller** [comment]
@InMatrix Perhaps the outer limit for the scope of the samples (and the widget catalog), in terms of what you might call atoms and molecules, are the public widget classes. I've made a rough list of them, there are about 220:  [all_widget_classes.txt](https://github.com/flutter/flutter/files/990445/all_widget_classes.txt)

I don't think there's anything wrong with imagining that all of those widget classes, as well as the supporting primitives (things like Color, Font, TextStyle, etc), would appear in at least one sample app.  Naturally we'll want to prioritize the work, and maybe we'll never get to all of them.

>  I think the sample catalog needs to distinguish itself by hosting samples at the levels of molecules and organisms. For example, an infinite scrolling UI pattern is a good candidate for the sample catalog, while a ListView belongs to the widget catalog. For the atomics (aka, widgets) in Flutter, we should provide sample code and screenshots in their API doc since this is where developers are accustomed to.

I mostly agree with this. In terms of the chemistry analogy: samples are simple organisms that demonstrate how the atoms and molecules (I'd call things like Scaffold, Container, and AppBar molecules) provided by Flutter can be used. How they fit together.

>  For example, an infinite scrolling UI pattern is a good candidate for the sample catalog, while a ListView belongs to the widget catalog.

I would like the ListView API doc and the ListView widget catalog entry to link to samples that highlight ListView. The links might point to every sample that uses ListView, or every sample with an annotation that indicates that it's intended to demonstrate ListView.  I'm fine with the sample catalog itself being organized around abstract themes like "infinite scrolling". I imagine that developers trying to find something in the sample catalog will mostly scan around for a likely screenshot.

**Comment by HansMuller** [comment]
@sethladd

> As you're thinking about scope, please include how we'll cross-link these samples into our Widget Catalog. It should be useful to bounce between Sample Catalog and Widget Catalog.

I think it would be useful for each widget catalog entry to have a single "sample applications" link that brought up the sample catalog page with a search query that produced (only) the samples marked as demonstrating that widget.

> I'd also love to see if/how we can display the source code for the sample, in the browser. Maybe it's just linking over to GitHub? If there's concern about bouncing the user back and forth, we might want to find a way to inline the source on the flutter.io/samples/____ pages.

In the fullness of time, I hope we'll link to a "Flutter Fiddle" editor where the developer could quickly try the code. For starters, just linking to GitHub would be OK.

> Did we think about how to more clearly mark the code in the sample that is specific to the sample (like, for instance, this Choice class), and the code that is generic and meant to be copy-pasted by the developer?

We could try marking the code that's the sample's focus with an annotation. Maybe @sample. If we're displaying the code with github we could generate a link like https://github.com/flutter/flutter/blob/master/examples/catalog/lib/animated_list.dart#L198-L202.

In addition to highlighting code that you might want to copy, this would help the reader understand where to start reading.

I suppose we could improve sample presentation further by generating a page rather than just linking to the source code on github. However I think there's more upside in providing a flutter fiddle app, and tuning the sample code for that.

**Comment by sethladd** [comment]
We met today and we have some ideas that we want to test. lgtm to land this, so we can start on writing the accompanying markdown docs.

**Pull request merged**